crates/wal-protocol/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use restate_types::identifiers::{LeaderEpoch, PartitionId, PartitionKey};
+use restate_types::invocation::{InvocationResponse, InvocationTermination, ServiceInvocation};
+use restate_types::message::MessageIndex;
+use restate_types::nodes_config::ConfigVersion;
+use restate_types::state_mut::ExternalStateMutation;
+
+use crate::control::AnnounceLeader;
+use crate::effects::BuiltinServiceEffects;
+use crate::timer::TimerValue;
+use restate_types::{GenerationalNodeId, PlainNodeId};
+
+pub mod control;
+pub mod effects;
+pub mod timer;
+
+/// The primary envelope for all messages in the system.
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Envelope {
+    pub header: Header,
+    pub command: Command,
+}
+
+/// Header is set on every message
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Header {
+    pub source: Source,
+    pub dest: Destination,
+    pub ack_mode: AckMode,
+}
+
+/// Identifies the source of a message
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Source {
+    /// Message is sent from another partition processor
+    Processor {
+        /// if possible, this is used to reroute responses in case of splits/merges
+        partition_id: PartitionId,
+        partition_key: Option<PartitionKey>,
+        /// The current epoch of the partition leader. Readers should observe this to decide which
+        /// messages to accept. Readers should ignore messages coming from
+        /// epochs lower than the max observed for a given partition id.
+        leader_epoch: LeaderEpoch,
+        /// This represents the logical clock of the sender at the time of writing this command.
+        /// The receiver can (optionally) use this message to detect duplication.
+        sequence_number: Option<MessageIndex>, // -- or sender_sequencer_number
+        /// Which node is this message from?
+        node_id: PlainNodeId,
+    },
+    /// Message is sent from an ingress node
+    Ingress {
+        /// The identity of the sender node. Generational for fencing. Ingress is
+        /// stateless, so we shouldn't respond to requests from older generation
+        /// if a new generation is alive.
+        node_id: GenerationalNodeId,
+        /// Last config version observed by sender. If this is a newer generation
+        /// or an unknown ID, we might need to update our config.
+        nodes_config_version: ConfigVersion,
+        // If unset and AckMode is dedup, we use the node_id as a key.
```

Comment:
With `node_id` being a `GenerationalNodeId` we will no longer have a stable dedup key. Right now this should not be a problem because the ingress does not require deduplication of its messages and the Kafka ingress will set a `dedup_key`.

Reply:
That's up to the use-case. If needed, dedup can be on `node_id.as_plain()` if we want the stable ID.